File: skills/generating-mermaid-diagrams/scripts/templates.js
/**
 * Mermaid Diagram Templates
 * Based on life-mermaid-master templates with syntax improvements from mermaid-visualizer
 */

export const DIAGRAM_TYPES = {
  FLOWCHART: 'flowchart',
  SEQUENCE: 'sequence',
  CLASS: 'class',
  STATE: 'state',
  ER: 'er',
  GANTT: 'gantt',
  PIE: 'pie',
  MINDMAP: 'mindmap',
  JOURNEY: 'journey',
};

export const COLOR_SCHEMES = {
  blue: {
    primary: '#3b82f6',
    secondary: '#1d4ed8',
    accent: '#60a5fa',
    bg_light: '#eff6ff',
    bg_dark: '#1e3a5f',
  },
  green: {
    primary: '#10b981',
    secondary: '#059669',
    accent: '#34d399',
    bg_light: '#d1fae5',
    bg_dark: '#064e3b',
  },
  purple: {
    primary: '#8b5cf6',
    secondary: '#7c3aed',
    accent: '#a78bfa',
    bg_light: '#ede9fe',
    bg_dark: '#4c1d95',
  },
  orange: {
    primary: '#f97316',
    secondary: '#ea580c',
    accent: '#fb923c',
    bg_light: '#ffedd5',
    bg_dark: '#7c2d12',
  },
};

export const ASPECT_RATIOS = {
  '4:3': { width: 800, height: 600 },
  '3:4': { width: 600, height: 800 },
  '16:9': { width: 960, height: 540 },
  '1:1': { width: 600, height: 600 },
  '3:2': { width: 750, height: 500 },
  '2:3': { width: 500, height: 750 },
};

export const FLOW_DIRECTIONS = {
  TB: 'TB',
  BT: 'BT',
  LR: 'LR',
  RL: 'RL',
};

// Semantic colors from mermaid-visualizer
export const SEMANTIC_COLORS = {
  input: { fill: '#d3f9d8', stroke: '#2f9e44' },
  planning: { fill: '#ffe3e3', stroke: '#c92a2a' },
  processing: { fill: '#e5dbff', stroke: '#5f3dc4' },
  action: { fill: '#ffe8cc', stroke: '#d9480f' },
  output: { fill: '#c5f6fa', stroke: '#0c8599' },
  storage: { fill: '#fff4e6', stroke: '#e67700' },
  learning: { fill: '#f3d9fa', stroke: '#862e9c' },
  metadata: { fill: '#e7f5ff', stroke: '#1971c2' },
  neutral: { fill: '#f8f9fa', stroke: '#868e96' },
};

export function generateFlowchart({ title, steps, branches, colors, direction = 'LR' }) {
  const lines = [`graph ${direction}`];

  // Escape node text to avoid syntax issues (from mermaid-visualizer)
  const escapeText = (text) => {
    let escaped = text.replace(/"/g, '『』');
    if (/\d\.\s/.test(escaped)) {
      escaped = escaped.replace(/(\d)\.\s/, '$1.');
    }
    return escaped;
  };

  // Generate nodes
  steps.forEach((step, i) => {
    lines.push(`    ${i}[${escapeText(step)}]`);
  });

  // Generate chain edges
  for (let i = 0; i < steps.length - 1; i++) {
    lines.push(`    ${i} --> ${i + 1}`);
  }

  // Generate branches
  if (branches) {
    Object.entries(branches).forEach(([condition, branchSteps]) => {
      const lastIdx = steps.length - 1;
      lines.push(`    ${lastIdx} -->|${condition}| ${lastIdx}_yes[${branchSteps[0]}]`);
      if (branchSteps.length > 1) {
        lines.push(`    ${lastIdx}_yes --> ${lastIdx}_done[${branchSteps[1]}]`);
      }
    });
  }

  return lines.join('\n');
}

export function generateSequence({ title, participants, messages, colors }) {
  const lines = [
    'sequenceDiagram',
    `    title ${title}`,
  ];

  participants.forEach(p => {
    const id = p.replace(/\s/g, '');
    lines.push(`    participant ${id}[${p}]`);
  });

  messages.forEach(m => {
    const fromId = m.from.replace(/\s/g, '');
    const toId = m.to.replace(/\s/g, '');
    const arrow = m.arrow || '->>';
    lines.push(`    ${fromId}${arrow}${toId}: ${m.text}`);
  });

  return lines.join('\n');
}

export function generateClassDiagram({ title, classes, relationships, colors }) {
  const lines = ['classDiagram', `    title ${title}`];

  classes.forEach(cls => {
    lines.push(`    class ${cls.name} {`);
    cls.attributes.forEach(attr => {
      lines.push(`        ${attr}`);
    });
    cls.methods.forEach(method => {
      lines.push(`        ${method}`);
    });
    lines.push('    }');
  });

  relationships.forEach(rel => {
    lines.push(`    ${rel.from} ${rel.type} ${rel.to} : ${rel.label}`);
  });

  return lines.join('\n');
}

export function generateGantt({ title, tasks, colors }) {
  const lines = [
    'gantt',
    `    title ${title}`,
    '    dateFormat YYYY-MM-DD',
    '    axisFormat %m-%d',
  ];

  tasks.forEach(task => {
    lines.push(`    ${task.name} :${task.section}, ${task.start}, ${task.duration}`);
  });

  return lines.join('\n');
}

export function generatePie({ title, data, colors }) {
  const lines = [
    'pie',
    `    title ${title}`,
  ];

  Object.entries(data).forEach(([label, value]) => {
    lines.push(`    "${label}" : ${value}`);
  });

  return lines.join('\n');
}

export function generateMindmap({ root, nodes, colors }) {
  const lines = ['mindmap', `    root(( ${root} ))`];

  const addNode = (parent, child, level = 1) => {
    const indent = ' '.repeat(level * 4);
    lines.push(`${indent}${child}`);
  };

  nodes.forEach(node => {
    if (node.parent) {
      const parentIndent = ' '.repeat(node.level * 4);
      lines.push(`${parentIndent}${node.parent}`);
    }
    addNode(node.parent || root, node.text, node.level || 1);
  });

  return lines.join('\n');
}

export const TEMPLATES = {
  'flowchart-basic': {
    type: DIAGRAM_TYPES.FLOWCHART,
    title: '基础流程图',
    generator: generateFlowchart,
    params: {
      steps: ['开始', '处理步骤1', '条件判断', '处理步骤2', '结束'],
      branches: { '是': ['分支处理A'], '否': ['分支处理B'] },
    },
  },
  'flowchart-decision': {
    type: DIAGRAM_TYPES.FLOWCHART,
    title: '决策流程',
    generator: generateFlowchart,
    params: {
      steps: ['问题', '收集信息', '分析选项', '做出决策', '执行'],
      branches: { '选项A': ['评估A'], '选项B': ['评估B'] },
    },
  },
  'sequence-api': {
    type: DIAGRAM_TYPES.SEQUENCE,
    title: 'API 调用时序',
    generator: generateSequence,
    params: {
      participants: ['客户端', 'API网关', '认证服务', '业务服务', '数据库'],
      messages: [
        { from: '客户端', to: 'API网关', text: 'HTTP请求', arrow: '->>' },
        { from: 'API网关', to: '认证服务', text: '验证Token', arrow: '->>' },
        { from: '认证服务', to: 'API网关', text: '验证结果', arrow: '-->>' },
        { from: 'API网关', to: '业务服务', text: '转发请求', arrow: '->>' },
        { from: '业务服务', to: '数据库', text: '数据操作', arrow: '->>' },
        { from: '数据库', to: '业务服务', text: '查询结果', arrow: '-->>' },
      ],
    },
  },
  'class-simple': {
    type: DIAGRAM_TYPES.CLASS,
    title: '简单类图',
    generator: generateClassDiagram,
    params: {
      classes: [
        { name: 'Animal', attributes: ['- name: String', '- age: int'], methods: ['+ speak(): void'] },
        { name: 'Dog', attributes: ['- breed: String'], methods: ['+ speak(): void'] },
        { name: 'Cat', attributes: ['- color: String'], methods: ['+ speak(): void'] },
      ],
      relationships: [
        { from: 'Dog', to: 'Animal', type: '--|>', label: '继承' },
        { from: 'Cat', to: 'Animal', type: '--|>', label: '继承' },
      ],
    },
  },
  'gantt-project': {
    type: DIAGRAM_TYPES.GANTT,
    title: '项目计划',
    generator: generateGantt,
    params: {
      tasks: [
        { name: '需求分析', section: '第一阶段', start: '2024-01-01', duration: '5d' },
        { name: '设计阶段', section: '第一阶段', start: 'after需求分析', duration: '7d' },
        { name: '开发实现', section: '第二阶段', start: 'after设计阶段', duration: '14d' },
        { name: '测试验证', section: '第二阶段', start: 'after开发实现', duration: '7d' },
        { name: '部署上线', section: '第三阶段', start: 'after测试验证', duration: '3d' },
      ],
    },
  },
  'pie-distribution': {
    type: DIAGRAM_TYPES.PIE,
    title: '资源分布',
    generator: generatePie,
    params: {
      data: { '核心功能': 40, '增强功能': 25, '维护更新': 20, '新特性': 15 },
    },
  },
};

File: skills/life-mermaid-master/scripts/templates.py
#!/usr/bin/env python3
"""
Mermaid Diagram Templates
Professional templates for different diagram types with validated syntax.
"""
from dataclasses import dataclass
from typing import List, Dict, Optional
from enum import Enum

class DiagramType(Enum):
    FLOWCHART = "flowchart"
    SEQUENCE = "sequenceDiagram"
    CLASS = "classDiagram"
    STATE = "stateDiagram-v2"
    ER = "er"
    GANTT = "gantt"
    PIE = "pie"
    MINDMAP = "mindmap"
    JOURNEY = "journey"

@dataclass
class NodeStyle:
    fill: str
    stroke: str
    stroke_width: int = 2
    rx: int = 5
    ry: int = 5

@dataclass
class DiagramTemplate:
    diagram_type: DiagramType
    nodes: List[str]
    edges: List[str]
    subgraphs: Dict[str, List[str]]
    styles: Dict[str, NodeStyle]
    direction: str = "LR"

# Professional color schemes
COLOR_SCHEMES = {
    "blue": {
        "primary": "#3b82f6",
        "secondary": "#1d4ed8",
        "accent": "#60a5fa",
        "bg_light": "#eff6ff",
        "bg_dark": "#1e3a8a",
        "success": "#10b981",
        "warning": "#f59e0b",
        "danger": "#ef4444",
    },
    "green": {
        "primary": "#10b981",
        "secondary": "#059669",
        "accent": "#34d399",
        "bg_light": "#ecfdf5",
        "bg_dark": "#064e3b",
        "success": "#10b981",
        "warning": "#f59e0b",
        "danger": "#ef4444",
    },
    "purple": {
        "primary": "#8b5cf6",
        "secondary": "#7c3aed",
        "accent": "#a78bfa",
        "bg_light": "#f5f3ff",
        "bg_dark": "#4c1d95",
        "success": "#10b981",
        "warning": "#f59e0b",
        "danger": "#ef4444",
    },
    "orange": {
        "primary": "#f97316",
        "secondary": "#ea580c",
        "accent": "#fb923c",
        "bg_light": "#fff7ed",
        "bg_dark": "#7c2d12",
        "success": "#10b981",
        "warning": "#f59e0b",
        "danger": "#ef4444",
    },
}

ASPECT_RATIOS = {
    "4:3": {"width": 800, "height": 600},
    "3:4": {"width": 600, "height": 800},
    "16:9": {"width": 960, "height": 540},
    "9:16": {"width": 540, "height": 960},
    "1:1": {"width": 600, "height": 600},
    "3:2": {"width": 750, "height": 500},
    "2:3": {"width": 500, "height": 750},
}

def generate_flowchart(
    title: str,
    steps: List[str],
    branches: Optional[Dict[str, List[str]]] = None,
    colors: Optional[Dict] = None,
    direction: str = "LR"
) -> str:
    """Generate flowchart Mermaid code"""
    if colors is None:
        colors = COLOR_SCHEMES["blue"]

    lines = [f"graph {direction}"]

    for i, step in enumerate(steps):
        node_id = chr(65 + i)
        lines.append(f"{node_id}[{step}]")

    # Connections
    for i in range(len(steps) - 1):
        lines.append(f"{chr(65 + i)} --> {chr(66 + i)}")

    # Branches
    if branches:
        branch_num = 0
        for step_idx, (condition, branch_steps) in enumerate(branches.items()):
            lines.append(f"{chr(65 + step_idx)} -->|{condition}| B{branch_num}[分支]")
            for j, bstep in enumerate(branch_steps):
                lines.append(f"B{branch_num}{chr(97 + j)}[{bstep}]")
            lines.append(f"B{branch_num} --> B{branch_num + 1}[汇聚]")
            branch_num += 1

    # Styles
    for i, step in enumerate(steps):
        node_id = chr(65 + i)
        fill = colors["accent"] if i == 0 else (colors["bg_light"] if i % 2 == 0 else colors["bg_light"])
        stroke = colors["secondary"]
        lines.append(f'style {node_id} fill:{fill},stroke:{stroke},stroke-width:2px')

    lines.append(f'linkStyle default stroke:{colors["primary"]},stroke-width:2px')

    return "\n    ".join(lines)

def generate_sequence(
    title: str,
    participants: List[str],
    messages: List[Dict],
    colors: Optional[Dict] = None
) -> str:
    """Generate sequence diagram Mermaid code"""
    if colors is None:
        colors = COLOR_SCHEMES["blue"]

    lines = [f"title {title}", ""]

    for p in participants:
        lines.append(f"participant {p}")

    lines.append("")
    for msg in messages:
        from_p = msg.get("from", "")
        to_p = msg.get("to", "")
        text = msg.get("text", "")
        arrow = msg.get("arrow", "-->")
        note = msg.get("note", "")

        if note:
            lines.append(f"Note over {from_p}: {note}")
        lines.append(f"{from_p} {arrow} {to_p}: {text}")

    return "sequenceDiagram\n    " + "\n    ".join(lines)

def generate_class_diagram(
    title: str,
    classes: List[Dict],
    relationships: List[Dict],
    colors: Optional[Dict] = None
) -> str:
    """Generate class diagram Mermaid code"""
    if colors is None:
        colors = COLOR_SCHEMES["purple"]

    lines = [f"title {title}", ""]

    for cls in classes:
        name = cls["name"]
        attrs = cls.get("attributes", [])
        methods = cls.get("methods", [])

        lines.append(f"class {name} {{")
        for attr in attrs:
            lines.append(f"    {attr}")
        for method in methods:
            lines.append(f"    {method}")
        lines.append("}")
        lines.append(f'style {name} fill:{colors["bg_light"]},stroke:{colors["primary"]},stroke-width:2px')

    lines.append("")
    for rel in relationships:
        from_cls = rel["from"]
        to_cls = rel["to"]
        type_ = rel.get("type", "--")
        label = rel.get("label", "")
        lines.append(f"{from_cls} {type_} {to_cls} : {label}")

    return "classDiagram\n    " + "\n    ".join(lines)

def generate_gantt(
    title: str,
    tasks: List[Dict],
    colors: Optional[Dict] = None
) -> str:
    """Generate gantt chart Mermaid code"""
    if colors is None:
        colors = COLOR_SCHEMES["green"]

    lines = [f"title {title}", ""]
    lines.append("dateFormat YYYY-MM-DD")
    lines.append("axisFormat %m-%d")

    sections = {}
    for task in tasks:
        section = task.get("section", "默认")
        if section not in sections:
            sections[section] = []
        sections[section].append(task)

    for section, task_list in sections.items():
        lines.append(f"section {section}")
        for task in task_list:
            name = task["name"]
            start = task.get("start", "")
            end = task.get("end", "")
            duration = task.get("duration", "")
            status = task.get("status", "active")

            if duration:
                lines.append(f"{name} :{status}, {start}, {duration}")
            else:
                lines.append(f"{name} :{status}, {start}, {end}")

    return "gantt\n    " + "\n    ".join(lines)

def generate_pie(
    title: str,
    data: Dict[str, float],
    colors: Optional[Dict] = None
) -> str:
    """Generate pie chart Mermaid code"""
    lines = [f'title {title}', ""]
    lines.append("show data")

    for label, value in data.items():
        # Escape special characters in labels
        safe_label = label.replace('"', '\\"')
        lines.append(f'    "{safe_label}" : {value}')

    return "pie\n    " + "\n    ".join(lines)

def generate_mindmap(
    title: str,
    nodes: List[str],
    children: Optional[Dict[str, List[str]]] = None,
    colors: Optional[Dict] = None
) -> str:
    """Generate mindmap Mermaid code"""
    if colors is None:
        colors = COLOR_SCHEMES["orange"]

    lines = [f"mindmap", f"  root(({title}))"]

    for node in nodes:
        lines.append(f"    {node}")

    if children:
        for parent, child_list in children.items():
            for child in child_list:
                lines.append(f"      {parent}")
                lines.append(f"        {child}")

    return "\n".join(lines)

def generate_er(
    title: str,
    entities: List[Dict],
    relationships: List[Dict],
    colors: Optional[Dict] = None
) -> str:
    """Generate ER diagram Mermaid code"""
    if colors is None:
        colors = COLOR_SCHEMES["blue"]

    lines = [f"er", f"    title {title}", ""]

    for entity in entities:
        name = entity["name"]
        attrs = entity.get("attributes", [])
        lines.append(f"    {name} {{")
        for attr in attrs:
            attr_type = attr.get("type", "int")
            attr_name = attr.get("name", "")
            lines.append(f"        {attr_type} {attr_name}")
        lines.append("    }")

    lines.append("")
    for rel in relationships:
        from_ent = rel["from"]
        to_ent = rel["to"]
        type_ = rel.get("type", "}")
        label = rel.get("label", "")
        lines.append(f"    {from_ent} {type_}--{to_ent} \"{label}\"")

    return "\n".join(lines)
